qutip/solver/nm_mcsolve.py

```diff
@@ -0,0 +1,148 @@
+__all__ = ['NonMarkovianMCSolver']
+
+from functools import partial
+import numpy as np
+import scipy
+from .mcsolve import MCSolver
+from .result import McResult
+from ..core import Qobj, QobjEvo, isket, ket2dm
+
+
+class NonMarkovianMCSolver(MCSolver):
+    r"""
+    Monte-Carlo solver for master equation with negative rates.
+    Based on the methods explained in arXiv:2209.08958 [quant-ph]
+    """
+    name = "nm_mcsolve"
+    resultclass = McResult
+    solver_options = {
+        "progress_bar": "text",
+        "progress_kwargs": {"chunk_size": 10},
+        "store_final_state": False,
+        "store_states": None,
+        "keep_runs_results": False,
+        "method": "adams",
+        "map": "serial",
+        "job_timeout": None,
+        "num_cpus": None,
+        "bitgenerator": None,
+        "mc_corr_eps": 1e-10,
+        "norm_steps": 5,
+        "norm_t_tol": 1e-6,
+        "norm_tol": 1e-4,
+        "completeness_rtol": None,
+        "completeness_atol": None,
+    }
+
+    # ops_and_rates is a list of tuples (L_i, Gamma_i), where Gamma_i = Gamma_i(t) is callable
+    def __init__(self, H, ops_and_rates, *args, options=None, **kwargs):
+        self.ops_and_rates = list(ops_and_rates)
+        self._mu_c = None
+
+        self._a_parameter, L = self._check_completeness(ops_and_rates, options)
+        if L is not None:
+            self.ops_and_rates.append((L, float))  # float() is zero
+
+        c_ops = self._compute_paired_c_ops()
+        super().__init__(H, c_ops, *args, options=options, **kwargs)
+
+    # Check whether op = sum(Li.dag() * Li) is proportional to identity
+    # If not, creates an extra Lindblad operator so that it is
+    # Returns: * the proportionality factor a
+    #          * the extra Lindblad operator (or None if not necessary)
+    @staticmethod
+    def _check_completeness(ops_and_rates, options=None):
+        tolerance_settings = {}
+        if options is not None:
+            rtol = options.get('completeness_rtol', None)
+            if rtol is not None:
+                tolerance_settings['rtol'] = rtol
+
+            atol = options.get('completeness_atol', None)
+            if atol is not None:
+                tolerance_settings['atol'] = atol
+
+        op = sum((f[0].dag() * f[0]).full() for f in ops_and_rates)
+
+        a_candidate = op[0, 0]
+        if np.allclose(a_candidate * np.eye(len(op)), op,
+                       **tolerance_settings):
+            return np.real(a_candidate), None
+
+        w, _ = np.linalg.eig(op)
+        a = max(np.real(w))
+        L = scipy.linalg.sqrtm(a * np.eye(len(op)) - op)  # new Lindblad operator
+        return a, Qobj(L)
```

Comment:
Here how it can be written using `Qobj`:
```suggestion
        op = sum((L.dag() * L) for L, _ in ops_and_rates)

        a_candidate = op.tr() / op.shape[0]
        with qt.CoreOptions(atol=self.options["completeness_atol"]):
            if op == a_candidate * qt.qeye(op.dims[0]):
                return np.real(a_candidate), None

        a = max(np.real(op.eigenenergies()))
        L = (a * qt.qeye(op.dims[0]) - op).sqrtm()
        return a, qt.Qobj(L)
```
This need to give a default float value to `completeness_atol` and do `self.options = options` in `__init__`.
(merge the latest changes in master if it raise an error.)